### DISCUSSION TITLE
disable live crossfade

### DISCUSSION THREAD

**Discussion by DAMaia1989** [AzuraCast/AzuraCast#6538] [Q&A]
hey guys
I'm new here and I'm still a little lost with the codes. I come to ask for your help to know if it is possible to do what I want.
I'm using the following code in LS but I want it not to crossfade when a dj starts live.

```
`def live_aware_crossfade(old, new) =
    if !to_live then
        # If going to the live show, play a simple sequence
        sequence([fade.out(old.source),fade.in(new.source)])
    else
        # Otherwise, use the smart transition
        cross.smart(old, new, fade_in=0.00, fade_out=2.00)
    end
end

radio = cross(minimum=1.0, duration=1.00, live_aware_crossfade, radio)`
```

is it possible to keep this crossfade only in autodj?

**Comment by Moonbase59** [comment]
Looking at the docs for [sequence](https://www.liquidsoap.info/doc-dev/reference.html#sequence), should be no problem. Try

```
        # If going to the live show, play a simple sequence
        sequence([old.source,new.source])
```

Couldn’t just now try it, it _might_ be:

```
        # If going to the live show, play a simple sequence
        sequence([old,new])
```

And remember that with newer releases (rolling!), you will have to change the `!to_live` into `to_live()`—it won’t work otherwise.

**Reply by DAMaia1989** [reply]
Thanks for the quick response, but unfortunately I've already tried it and it didn't work.
Ended up removing the crossfade also in autodj.
therefore, is there any way to configure a different crossfade for each situation? that is, a crossfade for when autodj is on and thus avoid silence between tracks and when someone goes live, do not crossfade in that passage?

**Reply by Moonbase59** [reply]
That’s the original intent "live_aware_crossfade" was invented for, I believe—to handle DJs/Streamers differently from the "normal" AutoDJ.

I wonder why you say it doesn’t work, maybe because AzuraCast brings in the same code directly after the place where you put yours in the Liquidsoap config… Hm. Will have to try on my station one of the days, I’m rather sure it should work—a `sequence`without any `fade.out()` and `fade.in()` should just be that: A sequence. Meaning, the DJ being immediately on the stream.

**Reply by DAMaia1989** [reply]
Did you try it? I continued to try but I noticed something curious, Azuracast does not respect the configuration that I edit manually in liquidsoap, it only respects the configuration that I chose in the panel, that is, either smartcrossfade, or simple or nothing. what could it be?

**Reply by Moonbase59** [reply]
Hmm… I remember that I also fought that for a time, until relatively recently, maybe some weeks ago, when Buster removed unneeded song annotations (which override what you’ve set in crossfade, but were needed for some time because of a Liquidsoap bug). Are you on a recent rolling release?

Btw, I always use "simple" crossfade—the smart has problems of its own. Setting crossfade to None in station settings also doesn’t work (when using custom crossfade code).

**Reply by DAMaia1989** [reply]
I already left it in simple mode like you said. What happens is that when I define 2 second to crossfade for example, when the DJ goes live the first 2 seconds are not heard. How can I solve this?

**Comment by Moonbase59** [comment]
@DAMaia1989 You’re right, this is totally _not_ self-explanatory. I took a little time to replicate all that in my station, and made two screenshots for you, hope this will help!

**Note:** This will _only_ function with a very new Liquidsoap (2.2.x or newer; I used 2.3.0+git@3dbbdc994) and the latest rolling release of Azuracast (I used Rolling Release #[6c27dfb](https://github.com/AzuraCast/AzuraCast/commit/6c27dfbd8237c4364d6ea12c3c43536f3d7772ed) (2023-08-30 12:20) here).

![Nite Radio – Mozilla Firefox_029](https://github.com/AzuraCast/AzuraCast/assets/3706922/c11e2689-ece7-4dd7-a583-665d1510d46a)

![Nite Radio – Mozilla Firefox_030](https://github.com/AzuraCast/AzuraCast/assets/3706922/02644261-d95f-46be-843e-805135dcfeff)

**Note 2:** The 3s AutoDJ fade-out in above example will be taken off the beginning of the live stream. Feel free to also adjust that duration to (almost) zero, if you wish.

**Note 3:** Due to a Liquidsoap bug, the _first song’s metadata_ will not be shown on your station. The ideal workaround (for now) would be to leave above fade-out _in_, and instruct your DJs to send a 3s "dummy" song at the start of their transmission. This would ensure listeners don’t have to wait a long time (minutes, maybe!) until the first song change happens, before they see the DJ’s metadata. (And this is actually how we do it currently, because we wish our intro or show title _to be seen_.)

To save you the work of copying code from a screenshot, here’s the custom Liquidsoap crossfade code (and a little more, hint, hint…) again:

```ruby
# Be sure to have ReplayGain applied before crossing.
radio = amplify(1.,override="replaygain_track_gain",radio)

# If a song has <song> <silence> <bonus track>,
# blank.eat() will skip <silence> and play <bonus track>
#radio = blank.eat(radio, max_blank=5.0, threshold=-60.0)
# blank.skip() will skip to next song and NOT play <bonus track>.
# Defaults: max_blank=20.0 (seconds), threshold=-40.0 (dB)
radio = blank.skip(radio, max_blank=5.0, threshold=-80.0)

def live_aware_crossfade(old, new) =
    if to_live() then
        # If going to the live show, play a simple sequence
        # fade out AutoDJ, do (almost) not fade in streamer
        sequence([fade.out(duration=3.0,old.source),fade.in(duration=0.1,new.source)])
    else
        # Otherwise, use the simple transition
        log("Using custom crossfade")
        cross.simple(old.source, new.source, fade_in=0.00, fade_out=3.50)
    end
end

radio = cross(minimum=0., duration=4.00, width=2.00, live_aware_crossfade, radio)
```

Adapt as needed. Good luck, and happy learning!

**Reply by DAMaia1989** [reply]
Thank you very much for all the help you provided, I really have no words. The code is quite interesting, Now i can better understand how it works with your tips. but there is an important situation, with this configuration we continue to lose the beginning of the live show for 4 seconds

`radio = cross(minimum=0., duration=4.00, width=2.00, live_aware_crossfade, radio)`

Now if I remove these 4 second from durations, I no longer lose the start of the live show but this way the songs end up being very spaced out, in other words, it ends up canceling out everything that was done above.

i have Rolling Release #[e251c0e]

**Reply by Moonbase59** [reply]
Same rolling release I currently use, so we can compare, very good.
Unfortunately, I have a live syndication running, so I’ll be able to test only tomorrow.

As I understand it _(I might be wrong, LS and realtime streaming are complicated beasts!)_, in
```ruby
radio = cross(minimum=0., duration=4.00, width=2.00, live_aware_crossfade, radio)
```
the `duration` is just the "window" (taken from the buffers), in which the new track gets started (without losing anything from the new track). This is used for power-level dependent crossfades, to find the "best" transition point. (If you were here, I’d now jot down some fader curves on a piece of paper, hee hee…)

But in
```ruby
sequence([fade.out(duration=3.0,old.source),fade.in(duration=0.1,new.source)])
```
the new track (live DJ) is _already playing_, so that code _should_ lose us 3 seconds of the live stream, because it’s a `sequence` and we use `fade.out(duration=3.0,…` to fade out the song playing in AutoDJ (while the live already plays).

One might try
```ruby
sequence([fade.out(duration=0.1,old.source),fade.in(duration=0.1,new.source)])
```
or even
```ruby
sequence([old.source, new.source])
```
to avoid losing the fade-out seconds from the live stream.

Or _even_ (not for the faint-hearted!) a `cross.simple`instead of the `sequence`…

Let me know about your findings. I’ll certainly give this a try, hopefully tomorrow.
Again, maybe I’m totally mistaken and we should ask @toots or @smimram…

**Reply by DAMaia1989** [reply]
Thanks again for your patience. Regarding this part of the code, I had already tried it the two ways you said and keep losing those 4 seconds

**Reply by Moonbase59** [reply]
Gave it a quick check. You’re right, losing ~4s, even when making the sequence fades 0.1s.
Hm. I’m rather positive I tested all these situations, and _had_ a complete stream a few versions back. Maybe something in Liquidsoap changed again… Sorry.

**Reply by DAMaia1989** [reply]
And now what can we do? Is there any option you can explain to me to go back to a previous version of Liquidsoap?

**Reply by Moonbase59** [reply]
Hard to say. I will hope for some enlightenment in this case, and in the meantime tell my streamers to send a 4-5s silent "dummy track" as their first song. Being on a rolling release can sometimes be a bumpy road…

As far as I understand, AzuraCast has no concept of "rolling back", so that might be hard. Except if you have a recent full VM "bare-metal" backup. I do those from time to time before upgrading, when an update seems risky to me.

Maybe someone else can chime in with a good idea?

**Comment by toots** [comment]
Hi there. Let me know if there's something I can do to help. It looks like @Moonbase59 analysis is correct. To have source-specific transitions we usually recommend to add a specific metadata to that source, for instance `type = "autodj"` and use that in the transition function to return a transition specific to `autodj`.

**Reply by DAMaia1989** [reply]
hi @toots , so in this case the advice is to create a different transition for live dj source? how can i implement it with the code that @Moonbase59 gave me?

**Comment by BusterNeece** [comment]
Hey folks! So, I realize our latest few versions may have some issues that prevent some stations from operating properly. While we're working on getting those things fixed, I also wanted to provide a way for you to roll back to an earlier stable release for your station if you want to.

If you want to roll back to a known-working stable release (i.e. 0.18.5), you can do the following via SSH:
 - `./docker.sh update` to update to the latest Rolling Release version. When prompted to update the Utility script, enter Y for yes.
 - `./docker.sh rollback 0.18.5`

It's recommended, as always, to back things up before doing this (though we do automatically take a backup of the database before modifying it). This should bring you back to a fully working instance of the older version of AzuraCast.

**Reply by DAMaia1989** [reply]
hey @BusterNeece  thanks for your response. I did as you said and I will change it to version 0.18.5 and insert the following code

`def live_aware_crossfade(old, new) =
    if !to_live then
        # If going to the live show, play a simple sequence
        sequence([fade.out(duration=0.,old.source),fade.in(duration=0.1,new.source)])
    else
        # Otherwise, use the smart transition
        log("Using custom crossfade")
        cross.simple(old.source, new.source, fade_in=0.00, fade_out=2.50)
    end
end

radio = cross(minimum=0., duration=4.00, width=2.00, live_aware_crossfade, radio)`

but it doesn't respect the configuration, it just doesn't do any crossfade either in the songs or in the live dj.

Is there something you're not doing correctly because of this version? I left all the parameters as the @Moonbase59  explained to me above

**Comment by DAMaia1989** [comment]
@BusterNeece Please, any solution to not miss the first seconds of the live show without killing the autodj crossfade? I updated again to the latest rolling version

**Reply by BusterNeece** [reply]
It's not for me to solve, I'm afraid. This is a Liquidsoap issue so your best bet would be to check with them over at https://github.com/savonet/liquidsoap.

**Reply by DAMaia1989** [reply]
@toots Were you able to verify this situation? I noticed that some people have also reported this same problem with the LS

**Comment by DAMaia1989** [comment]
hello guys! I hope you are well :) I don't know if you reached any conclusion or solution. But I wanted to share with you what is happening to me now. I didn't change any settings or do anything and now azuracast doesn't cut the first seconds of the beginning. Now,  when the live show ends does it start a new song as it always does but you still hear 3 seconds of the song that was playing on the autodj initially . Are you also having this problem? What could it be?

**Reply by DAMaia1989** [reply]
I forgot to mention that I currently have the Rolling Release #0791831 (2023-10-01 5:33)
